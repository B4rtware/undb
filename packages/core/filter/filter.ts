import type { CompositeSpecification } from '@egodb/domain'
import type { Option } from 'oxide.ts'
import { None, Some } from 'oxide.ts'
import { z } from 'zod'
import type { ISelectFieldValue } from '../field/select-field.type'
import { selectFieldValue } from '../field/select-field.type'
import {
  DateEqual,
  DateGreaterThan,
  DateGreaterThanOrEqual,
  DateIsToday,
  DateLessThan,
  DateLessThanOrEqual,
  NumberEqual,
  NumberGreaterThan,
  NumberGreaterThanOrEqual,
  NumberLessThan,
  NumberLessThanOrEqual,
  SelectEqual,
  SelectIn,
  StringContain,
  StringEndsWith,
  StringEqual,
  StringStartsWith,
} from '../record'
import {
  $is_today,
  dateFilterOperators,
  numberFilterOperators,
  selectFilterOperators,
  stringFilterOperators,
} from './operators'

const baseFilter = z.object({
  path: z.string().min(1),
})

export const stringFilterValue = z.string().nullable()
const stringFilter = z
  .object({
    type: z.literal('string'),
    operator: stringFilterOperators,
    value: stringFilterValue,
  })
  .merge(baseFilter)

export type IStringFilter = z.infer<typeof stringFilter>
export type IStringFilterOperator = z.infer<typeof stringFilterOperators>

export const numberFilterValue = z.number().nullable()
const numberFilter = z
  .object({
    type: z.literal('number'),
    operator: numberFilterOperators,
    value: numberFilterValue,
  })
  .merge(baseFilter)
export type INumberFilter = z.infer<typeof numberFilter>
export type INumberFilterOperator = z.infer<typeof numberFilterOperators>

export const dateFilterValue = z.date().nullable().optional()
const dateFilter = z
  .object({
    type: z.literal('date'),
    operator: dateFilterOperators,
    value: dateFilterValue,
  })
  .merge(baseFilter)
export type IDateFilter = z.infer<typeof dateFilter>

export const selectFilterValue = selectFieldValue.or(selectFieldValue.array()).nullable()
export type ISelectFilterValue = z.infer<typeof selectFieldValue>
const selectFilter = z
  .object({
    type: z.literal('select'),
    operator: selectFilterOperators,
    value: selectFilterValue,
  })
  .merge(baseFilter)

export type ISelectFilter = z.infer<typeof selectFilter>

export const filterValue = z.union([stringFilterValue, numberFilterValue, dateFilterValue, selectFilterValue])
export type IFilterValue = z.infer<typeof filterValue>

export const operaotrs = z.union([
  stringFilterOperators,
  numberFilterOperators,
  dateFilterOperators,
  selectFilterOperators,
])
export type IOperator = z.infer<typeof operaotrs>

const filter = z.discriminatedUnion('type', [stringFilter, numberFilter, dateFilter, selectFilter])

const $and = z.literal('$and')
const $or = z.literal('$or')
const $not = z.literal('$not')

const conjunctions = z.union([$and, $or, $not])
export type IConjunction = z.infer<typeof conjunctions>

export type IFilter = z.infer<typeof filter>
export type IFilters = IFilter[]

export interface IGroup {
  conjunction: IConjunction
  children: IFilterOrGroupList
}

const group: z.ZodType<IGroup> = z.lazy(() =>
  z.object({
    conjunction: conjunctions,
    children: z.union([group, filter]).array().nonempty(),
  }),
)

const filterOrGroup = filter.or(group)
export type IFilterOrGroup = z.infer<typeof filterOrGroup>

export const filterOrGroupList = filterOrGroup.array()
export type IFilterOrGroupList = z.infer<typeof filterOrGroupList>
export const rootFilter = z.union([filterOrGroup, filterOrGroupList])
export type IRootFilter = z.infer<typeof rootFilter>

const isGroup = (filterOrGroup: IFilterOrGroup): filterOrGroup is IGroup => {
  return Object.hasOwn(filterOrGroup, 'conjunction')
}

const isFilter = (filterOrGroup: IFilterOrGroup): filterOrGroup is IFilter => {
  return Object.hasOwn(filterOrGroup, 'type') && Object.hasOwn(filterOrGroup, 'operator')
}

const convertStringFilter = (filter: IStringFilter): Option<CompositeSpecification> => {
  if (!filter.value) {
    return None
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(new StringEqual(filter.path, filter.value))
    }
    case '$neq': {
      return Some(new StringEqual(filter.path, filter.value).not())
    }
    case '$contains': {
      return Some(new StringContain(filter.path, filter.value))
    }
    case '$starts_with': {
      return Some(new StringStartsWith(filter.path, filter.value))
    }
    case '$ends_with': {
      return Some(new StringEndsWith(filter.path, filter.value))
    }

    default:
      return None
  }
}

const convertNumberFilter = (filter: INumberFilter): Option<CompositeSpecification> => {
  if (!filter.value) {
    return None
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(new NumberEqual(filter.path, filter.value))
    }
    case '$neq': {
      return Some(new NumberEqual(filter.path, filter.value).not())
    }
    case '$gt': {
      return Some(new NumberGreaterThan(filter.path, filter.value))
    }
    case '$gte': {
      return Some(new NumberGreaterThanOrEqual(filter.path, filter.value))
    }
    case '$lt': {
      return Some(new NumberLessThan(filter.path, filter.value))
    }
    case '$lte': {
      return Some(new NumberLessThanOrEqual(filter.path, filter.value))
    }
    default:
      return None
  }
}

const convertSelectFilter = (filter: ISelectFilter): Option<CompositeSpecification> => {
  if (!filter.value) {
    return None
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(new SelectEqual(filter.path, filter.value as ISelectFieldValue))
    }
    case '$neq': {
      return Some(new SelectEqual(filter.path, filter.value as ISelectFieldValue).not())
    }
    case '$in': {
      return Some(new SelectIn(filter.path, filter.value as ISelectFieldValue[]))
    }
    case '$nin': {
      return Some(new SelectIn(filter.path, filter.value as ISelectFieldValue[]).not())
    }

    default: {
      return None
    }
  }
}

const convertDateFilter = (filter: IDateFilter): Option<CompositeSpecification> => {
  if (filter.operator === $is_today.value) {
    return Some(new DateIsToday(filter.path))
  }

  if (!filter.value) {
    return None
  }

  switch (filter.operator) {
    case '$eq': {
      return Some(new DateEqual(filter.path, filter.value))
    }
    case '$neq': {
      return Some(new DateEqual(filter.path, filter.value).not())
    }
    case '$gt': {
      return Some(new DateGreaterThan(filter.path, filter.value))
    }
    case '$gte': {
      return Some(new DateGreaterThanOrEqual(filter.path, filter.value))
    }
    case '$lt': {
      return Some(new DateLessThan(filter.path, filter.value))
    }
    case '$lte': {
      return Some(new DateLessThanOrEqual(filter.path, filter.value))
    }
    default:
      return None
  }
}

const convertFilter = (filter: IFilter): Option<CompositeSpecification> => {
  switch (filter.type) {
    case 'string':
      return convertStringFilter(filter)
    case 'number':
      return convertNumberFilter(filter)
    case 'date':
      return convertDateFilter(filter)
    case 'select':
      return convertSelectFilter(filter)
    default:
      return None
  }
}

const convertFilterOrGroup = (filterOrGroup: IFilterOrGroup): Option<CompositeSpecification> => {
  if (isGroup(filterOrGroup)) {
    return convertFilterOrGroupList(filterOrGroup.children, filterOrGroup.conjunction)
  } else if (isFilter(filterOrGroup)) {
    return convertFilter(filterOrGroup)
  }

  return None
}

const convertFilterOrGroupList = (
  filterOrGroupList: IFilterOrGroupList,
  conjunction: IConjunction = '$and',
): Option<CompositeSpecification> => {
  let spec: Option<CompositeSpecification> = None
  for (const filter of filterOrGroupList) {
    if (spec.isNone()) {
      spec = convertFilterOrGroup(filter)
      if (conjunction === '$not') {
        return spec.map((s) => s.not())
      }
    } else {
      if (isFilter(filter)) {
        spec = spec.map((left) => {
          const right = convertFilterOrGroup(filter)
          if (right.isSome()) {
            if (conjunction === '$and') {
              return left.and(right.unwrap())
            } else if (conjunction === '$or') {
              return left.or(right.unwrap())
            }
            return left.and(right.unwrap().not())
          }
          return left
        })
      } else if (isGroup(filter)) {
        spec = convertFilterOrGroupList(filter.children, filter.conjunction)
      }
    }
  }

  return spec
}

export const convertFilterSpec = (filter: IRootFilter): Option<CompositeSpecification> => {
  if (Array.isArray(filter)) {
    return convertFilterOrGroupList(filter)
  }

  return convertFilterOrGroup(filter)
}
