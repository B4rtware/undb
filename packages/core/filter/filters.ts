import { ValueObject } from '@egodb/domain'
import type { IFilters } from './operators'

export class Filters extends ValueObject<IFilters> {
  get value() {
    return this.props
  }
}
