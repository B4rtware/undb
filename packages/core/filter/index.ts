export * from './filter'
export * from './root-filter'
