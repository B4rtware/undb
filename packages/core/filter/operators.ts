import { z } from 'zod'
import type { INumberFieldValue } from '../field/number-field.type'
import { numberFieldValue } from '../field/number-field.type'
import type { ITextFieldValue } from '../field/text-field.type'
import { textFieldValue } from '../field/text-field.type'

export type Filter<TSchema extends Record<string, ComparableFieldValue> = Record<string, ComparableFieldValue>> =
  | RootOperator
  | TSchema
  | Record<string, Operator>

export interface RootOperator {
  $and?: Filter[]
  $or?: Filter[]
}

type ComparableFieldValue = ITextFieldValue | INumberFieldValue

type FilterOperator = { $eq?: ComparableFieldValue } | { $neq?: ComparableFieldValue }

export type Operator = ComparableFieldValue | FilterOperator | { $not?: FilterOperator }

const $comparableFieldValue = z.union([textFieldValue, numberFieldValue])

const $eq = z.object({ $eq: $comparableFieldValue })
const $neq = z.object({ $neq: $comparableFieldValue })

export const $filterOperator = z.union([$eq, $neq])

const $not = z.lazy(() => z.object({ $not: $filterOperator }))

export const $filter: z.ZodType<Filter> = z.lazy(() =>
  z.union([z.record($comparableFieldValue.or($filterOperator).or($not)), $rootOperator]),
)

const $filters = z.array($filter).optional()
const $rootOperator = z.object({
  $and: $filters,
  $or: $filters,
})
