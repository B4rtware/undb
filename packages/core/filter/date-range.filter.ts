import { z } from 'zod'
import { dateRangeFieldValue } from '../field/date-range-field.type.js'
import { baseFilter } from './filter.base.js'
import { dateRangeFilterOperators } from './operators.js'

export const dateRangeFilterValue = dateRangeFieldValue
export const dateRangeFilter = z
  .object({
    type: z.literal('date-range'),
    operator: dateRangeFilterOperators,
    value: dateRangeFilterValue,
  })
  .merge(baseFilter)
export type IDateRangeFilter = z.infer<typeof dateRangeFilter>
