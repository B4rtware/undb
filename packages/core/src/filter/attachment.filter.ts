import { z } from 'zod'
import { baseFilter } from './filter.base.js'
import { attachmentFilterOperators } from './operators.js'

export const attachmentFilterValue = z.string().nullable()
export const attachmentFilter = z
  .object({
    type: z.literal('attachment'),
    operator: attachmentFilterOperators,
    value: attachmentFilterValue,
  })
  .merge(baseFilter)

export type IAttachmentFilter = z.infer<typeof attachmentFilter>
export type IAttachmentFilterOperator = z.infer<typeof attachmentFilterOperators>
