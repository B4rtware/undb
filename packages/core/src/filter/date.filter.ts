import { z } from 'zod'
import { baseFilter } from './filter.base.js'
import { dateFilterOperators } from './operators.js'

export const dateFilterValue = z.date().nullable().optional()
export const dateFilter = z
  .object({
    type: z.literal('date'),
    operator: dateFilterOperators,
    value: dateFilterValue,
  })
  .merge(baseFilter)
export type IDateFilter = z.infer<typeof dateFilter>
