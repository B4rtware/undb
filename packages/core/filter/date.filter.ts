import { z } from 'zod'
import { baseFilter } from './filter.base'
import { dateFilterOperators } from './operators'

export const dateFilterValue = z.date().nullable().optional()
export const dateFilter = z
  .object({
    type: z.literal('date'),
    operator: dateFilterOperators,
    value: dateFilterValue,
  })
  .merge(baseFilter)
export type IDateFilter = z.infer<typeof dateFilter>
