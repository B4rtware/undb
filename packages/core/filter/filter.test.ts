import { convertFilterSpec, IRootFilter, rootFilter } from './filter'

test.each<IRootFilter>([
  {
    type: 'string',
    operator: '$eq',
    path: 'name',
    value: 'hello',
  },
  {
    type: 'string',
    operator: '$contains',
    path: 'name',
    value: 'hello',
  },
  {
    type: 'string',
    operator: '$starts_with',
    path: 'name',
    value: 'starts with',
  },
  {
    type: 'string',
    operator: '$ends_with',
    path: 'name',
    value: 'ends with',
  },
  {
    type: 'number',
    operator: '$neq',
    path: 'name',
    value: 1,
  },
  {
    type: 'number',
    operator: '$neq',
    path: 'name.nested',
    value: 1,
  },
  [
    {
      conjunction: '$and',
      children: [
        {
          type: 'string',
          operator: '$eq',
          path: 'name',
          value: 'hello',
        },
      ],
    },
  ],
  [
    {
      conjunction: '$or',
      children: [
        {
          type: 'string',
          operator: '$eq',
          path: 'name',
          value: 'hello',
        },
      ],
    },
  ],
])('should create root filter', (filter) => {
  const parsed = rootFilter.parse(filter)
  expect(parsed).toEqual(filter)

  const spec = convertFilterSpec(filter)
  expect(spec).toMatchSnapshot()
})
