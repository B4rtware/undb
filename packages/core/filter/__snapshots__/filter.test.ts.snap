// Vitest Snapshot v1

exports[`should create root filter 1`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": "hello",
  },
}
`;

exports[`should create root filter 2`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringContain {
    "fieldId": "name",
    "value": "hello",
  },
}
`;

exports[`should create root filter 3`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringStartsWith {
    "fieldId": "name",
    "value": "starts with",
  },
}
`;

exports[`should create root filter 4`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEndsWith {
    "fieldId": "name",
    "value": "ends with",
  },
}
`;

exports[`should create root filter 5`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): Not {
    "spec": NumberEqual {
      "fieldId": "name",
      "value": 1,
    },
  },
}
`;

exports[`should create root filter 6`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): Not {
    "spec": NumberEqual {
      "fieldId": "name.nested",
      "value": 1,
    },
  },
}
`;

exports[`should create root filter 7`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": "hello",
  },
}
`;

exports[`should create root filter 8`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): StringEqual {
    "fieldId": "name",
    "value": "hello",
  },
}
`;

exports[`should create root filter 9`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThan {
    "fieldId": "gt",
    "value": 1,
  },
}
`;

exports[`should create root filter 10`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberGreaterThanOrEqual {
    "fieldId": "gte",
    "value": 1,
  },
}
`;

exports[`should create root filter 11`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThan {
    "fieldId": "lt",
    "value": 1,
  },
}
`;

exports[`should create root filter 12`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): NumberLessThanOrEqual {
    "fieldId": "lte",
    "value": 1,
  },
}
`;

exports[`should create root filter 13`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): DateIsToday {
    "fieldId": "is_today",
    "value": null,
  },
}
`;

exports[`should create root filter 14`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsTrue {
    "fieldId": "field",
    "value": true,
  },
}
`;

exports[`should create root filter 15`] = `
OptionType {
  Symbol(T): true,
  Symbol(Val): BoolIsFalse {
    "fieldId": "field",
    "value": false,
  },
}
`;
